### DISCUSSION TITLE
Include navigation state in request sent to loader

### DISCUSSION THREAD

**Discussion by RoBackman** [remix-run/react-router#11227] [Proposals]
When navigating to a page, one can currently add "state" to the navigation that then becomes available in useLocation().state

I would like to suggest an improvement to this by making it available in the loader function.

Browsed the underlying code a bit, and some of my thoughts following:
- In case state is set when navigating, the request that is forwarded to the loader could be of type "post" with the state included in the request body.
- As the history state spec also requires the passed data to be serializable -> would fit quite nicely json-serialized in the body.
- By using the native requests this would probably not break any SSR functionality in remix.
- This could in part be used to limit user from tampering with statefull things that would otherwise need to be included in the url (dates etc.).
- useLocation().state appears stable between refreshes and back/forward navigation -> possible to include it again when navigating forward/back

Sorry for the messy explanation, feel free to restructure it and ask if anything is unclear.

**Comment by kberridge** [comment]
We were hoping for exactly this too.  I have a scenario where a page has a loader request to load information about how the page should initially be rendered, but when a user clicks certain buttons on the page that navigate, we know we want to clear that state, so we do not need to load it.  It seemed like passing history state to `navigate` would be a great way to accomplish this, but we haven't found a way to access that state in the loader.  

Our workaround for now was to set and check global state on `window`.  We're fully client side rendered, no SSR, so I was hoping to find an optional historyState parameter to loader.

**Comment by rubber-duck-software** [comment]
Linking #11274 . 

A general ability to read/write history.state has tons of uses!

**Comment by RedVelocity** [comment]
I'm on Remix v2, was really hoping RR7 would finally add this feature. I guess we'll wait for Remix 3?

Edit: Adding a genuine use case of a weather app

- User searches for an address on client page, the page hits a geocoding api to get addresses/location based on search.
- The results are rendered as a Link element; user can click the Link to get the weather for selected location.
- The Link triggers a Loader which does the geocoding **again** on the route path address and then sends the weather data for the location.

If the Link can pass the location object as a state to loader, the app can completely **skip a second call** to geocoding api.

Of course, I can already include the geocoding result as Link params but that's not what I'm looking for. I would like my Links to human readable and shareable format.

**Reply by gustavopch** [reply]
You probably mean RR8.

**Reply by RedVelocity** [reply]
> You probably mean RR8.

No

**Comment by RedVelocity** [comment]
Hi, is there any update on this? I still believe this is a valid use case for navigation state.
